File: Introduction to Computer Science/Reference/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long card_number;
    int count = 0;
    int type = 0;

    card_number = get_long("Number: ");

    long card_number_copy = card_number;

    // counts number of digits
    while (card_number_copy > 0)
    {
        count++;
        card_number_copy /= 10;
    }

    if (count == 13)
    {

        if (card_number / 1000000000000 == 4)
        {
            type = 1;
        }
    }

    else if (count == 15)
    {
        int am15 = card_number / 10000000000000;
        if (am15 == 34 || am15 == 37)
        {
            type = 2;
        }
    }

    else if (count == 16)
    {
        int mc16 = card_number / 100000000000000;
        int visa16 = card_number / 1000000000000000;

        if (visa16 == 4)
        {
            type = 1;
        }
        else if (mc16 == 51 || mc16 == 52 || mc16 == 53 || mc16 == 54 || mc16 == 55)
        {
            type = 3;
        }
    }

    if (type > 0)
    {
        int last = 0, secToLast = 0, x;

        while (card_number > 0)
        {
            last += card_number % 10;

            card_number /= 10;

            x = (card_number % 10) * 2;

            if (x > 9)
            {
                int quotient = x / 10;
                int remainder = x % 10;
                secToLast += (quotient + remainder);
            }
            else
            {
                secToLast += x;
            }

            card_number /= 10;
        }
        if (((last + secToLast) % 10) == 0)
        {
            if (type == 1)
            {
                printf("VISA\n");
            }
            else if (type == 2)
            {
                printf("AMEX\n");
            }
            else if (type == 3)
            {
                printf("MASTERCARD\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Introduction to Computer Science/Week 6 - Python/credit/credit.py
from cs50 import get_int

import math


def main():
    # get input
    card = get_int("Number: ")

    result = 0
    type = ""

    # check master
    if (card >= 1000000000000000):
        if math.floor(card / 100000000000000) == 51 or math.floor(card / 100000000000000) == 52 or math.floor(card / 100000000000000) == 53 or math.floor(card / 100000000000000) == 54 or math.floor(card / 100000000000000) == 55:
            type = "MASTERCARD"
            result = checkSum(16, card)

        elif (math.floor(card / 1000000000000000) == 4):
            type = "VISA"
            result = checkSum(16, card)

    # check amex
    elif (card >= 100000000000000):

        if (math.floor(card / 10000000000000) == 34 or math.floor(card / 10000000000000) == 37):
            type = "AMEX"
            result = checkSum(15, card)

    # check visa
    elif (card >= 1000000000000):
        if (math.floor(card / 1000000000000) == 4):
            type = "VISA"
            result = checkSum(13, card)

    # check invalid
    else:
        type = "INVALID"

    # check card number is valid
    if (result == 1):
        print(type)
    else:
        type = "INVALID"
        print(type)


def checkSum(length, card):

    # declare program variables
    sec_to_last = 0
    last = 0
    divider = 1
    turn = 0

    # loop for length
    for i in range(length):
        c = math.floor(card / divider)
        remainder = c % 10

        if (turn == 0):
            last += remainder
            turn = 1

        elif (turn == 1):
            remainder = remainder * 2
            if (remainder < 10):
                sec_to_last += remainder
            else:
                sec_to_last += math.floor(remainder / 10)
                sec_to_last += remainder % 10

            turn = 0
        divider *= 10

    # check valid checksum
    if ((sec_to_last + last) % 10 == 0):
        return 1
    else:
        return 0


main()